### PR TITLE
CharCounterApplication

### PR DESCRIPTION
**Pull request by mikhailtarasevich**
The app count unique characters in a text and print them in the next view: "Character" = amountOfThisCharacterInText.

### CODE REVIEW

**Comment by mikhailtarasevich** [comment]
CharCounterCmdRun

Run the app in CMD

**Comment by mikhailtarasevich** [comment]
JaCoCoCharCounter

Unit tests' coverage

File: src/test/java/CharCounterApplicationTest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	
+	Validator validator = new ValidatorImpl();
+	CharCounter charCounter = new CharCounterImpl();
+	HashMapToStringConverter hashMapToStringConverter = new HashMapToStringConverterImpl();
+	CharCounterApplication �harCounterApplication = new CharCounterApplication(validator, charCounter, hashMapToStringConverter);
+	
+	
+	@Test
+	public void countCharactersInText_inputIsString_expectedStringWithAmountOfUniqueCharacter() {
+		assertThat(�harCounterApplication.countCharactersInText("He 22")).isEqualTo("He 22" + "\n" + "\"" + "H" + "\"" + " = 1" + "\n" + "\"" + "e" + "\"" + " = 1" + "\n" + "\"" + " " + "\"" + " = 1" + "\n" + "\"" + "2" + "\"" + " = 2");
```
Comment (AlexeyChrikov94) [inline src/test/java/CharCounterApplicationTest.java:24]: ñ ??
we use utf-8 encoding

File: pom.xml [hunk 2]
```diff
@@ -1,6 +1,122 @@
...
+
+	<properties>
+		<junit-jupiter.version>5.7.0-M1</junit-jupiter.version>
+		<org.hamcrest.version>2.2</org.hamcrest.version>
+		<assertj-core.version>3.22.0</assertj-core.version>
+		<mockito-core.version>2.28.2</mockito-core.version>
+		<surefire-junit4.version>2.22.0</surefire-junit4.version>
+		<maven-surefire-plugin.version>2.22.0</maven-surefire-plugin.version>
+		<maven-compiler-plugin.version>3.8.1</maven-compiler-plugin.version>
+
```
Comment (AlexeyChirkov94) [inline pom.xml]: remove empty line

File: src/test/java/CharCounterApplicationTest.java (hunk 1 above)
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: ñ ??
we use utf-8 encoding

File: pom.xml [hunk 3]
```diff
@@ -1,6 +1,122 @@
...
+			<artifactId>junit-jupiter</artifactId>
+			<version>${junit-jupiter.version}</version>
+			<scope>test</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>org.hamcrest</groupId>
+			<artifactId>hamcrest</artifactId>
+			<version>${org.hamcrest.version}</version>
+			<scope>test</scope>
```
Comment (AlexeyChirkov94) [inline pom.xml]: where you use hamcrest?

File: pom.xml [hunk 4]
```diff
@@ -1,6 +1,122 @@
...
+			<artifactId>mockito-core</artifactId>
+			<version>${mockito-core.version}</version>
+			<scope>test</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.maven.surefire</groupId>
+			<artifactId>surefire-junit4</artifactId>
+			<version>${surefire-junit4.version}</version>
+		</dependency>
```
Comment (AlexeyChirkov94) [inline pom.xml:44]: you needn't add separate dependency. Plugin will enough

File: pom.xml [hunk 5]
```diff
@@ -1,6 +1,122 @@
...
+
+	<build>
+		<plugins>
+
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-surefire-plugin</artifactId>
+				<version>${maven-surefire-plugin.version}</version>
+				<configuration>
+					<testFailureIgnore>true</testFailureIgnore>
```
Comment (AlexeyChirkov94) [inline pom.xml]: testFailureIgnore = true ?

File: pom.xml [hunk 6]
```diff
@@ -1,6 +1,122 @@
...
+					</execution>
+					<execution>
+						<id>report</id>
+						<phase>test</phase>
+						<goals>
+							<goal>report</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
```
Comment (AlexeyChirkov94) [inline pom.xml:106]: exclude ConsoleForCharCounterApplication from jacoco check

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImpl.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+import java.util.LinkedHashMap;
+
+public class CharCounterImpl implements CharCounter {
+
+	public LinkedHashMap<Character, Integer> uniqueCharCounter (String text) {
+		char[] strToArray = text.toCharArray();
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImpl.java]: use stream api here

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImpl.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+import java.util.LinkedHashMap;
+
+public class HashMapToStringConverterImpl implements HashMapToStringConverter {
+
+	public String convertHashMapToString(LinkedHashMap<Character, Integer> map, String text) {
+	    StringBuilder mapAsString = new StringBuilder(text);
+	    for (Character key : map.keySet()) {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImpl.java]: use stream api here

File: src/test/java/CharCounterApplicationTest.java (hunk 1 above)
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: why not merge into one string?

File: src/test/java/CharCounterApplicationTest.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	
+	@Test
+	public void countCharactersInText_inputIsString_expectedStringWithAmountOfUniqueCharacter() {
+		assertThat(�harCounterApplication.countCharactersInText("He 22")).isEqualTo("He 22" + "\n" + "\"" + "H" + "\"" + " = 1" + "\n" + "\"" + "e" + "\"" + " = 1" + "\n" + "\"" + " " + "\"" + " = 1" + "\n" + "\"" + "2" + "\"" + " = 2"); 
+	}
+	
+	@Test
+	public void countCharactersInText_inputIsNull_expectedException() {
+		Assertions.assertThrows(IllegalArgumentException.class, () -> {
+			�harCounterApplication.countCharactersInText(null);
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: same about utf-8

File: src/test/java/CharCounterApplicationTest.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import com.mikhail.tarasevich.charcounter.validator.Validator;
+import com.mikhail.tarasevich.charcounter.validator.ValidatorImpl;
+
+class CharCounterApplicationTest {
+	
+	Validator validator = new ValidatorImpl();
+	CharCounter charCounter = new CharCounterImpl();
+	HashMapToStringConverter hashMapToStringConverter = new HashMapToStringConverterImpl();
+	CharCounterApplication �harCounterApplication = new CharCounterApplication(validator, charCounter, hashMapToStringConverter);
+
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: You must use mock of object at this test, 
please use Mockito here
https://www.baeldung.com/mockito-annotations - about mock injects
https://www.baeldung.com/mockito-junit-5-extension - using with junit 5 
https://www.baeldung.com/mockito-series - all tutarials
https://habr.com/ru/post/444982/ - in Russian

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: we don`t use wildcard imports

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		assertEquals(expectedLinkedHashMap, charCounterImpl.uniqueCharCounter(text));
+		
+	}
+	
+	@Test
+	void uniqueCharCounter_inputIsNull_expectedNullPointerException() {
+		
+		expectedLinkedHashMap.put(null, 1);
+		text = null;
+		Assertions.assertThrows(NullPointerException.class, () -> {
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: why not use static import of method ?

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java]: the same

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	@Test
+	void convertHashMapToString_inputHashMapWithAmountOfUniqueCharacter_expectedStringWithAmountOfUniqueCharacter() {
+		inputString = "Hello 22";
+		inputHashMap.put('H', 1);
+		inputHashMap.put('e', 1);
+		inputHashMap.put('l', 2);
+		inputHashMap.put('o', 1);
+		inputHashMap.put(' ', 1);
+		inputHashMap.put('2', 2);		
+		expectedString = inputString + "\n" + "\"" + "H" + "\"" + " = 1" + "\n" + "\"" + "e" + "\"" + " = 1" + "\n" + "\"" + "l" + "\"" + " = 2" + "\n" + "\"" + "o" + "\"" + " = 1" + "\n" + "\"" + " " + "\"" + " = 1" + "\n" + "\"" + "2" + "\"" + " = 2";
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java]: the same? why not in one string? why separate?

File: src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.jupiter.api.Test;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+class ValidatorImplTest {
+
+	ValidatorImpl validatorTest = new ValidatorImpl();
+
+	@Test
+	void validate_InputWhiteSpaces_ExpectedIllegalArgumentException() {
+		Assertions.assertThrows(IllegalArgumentException.class, () -> {
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java]: please use static import here and below
and we use assertj librarty for test, here you use jupiter.api.Assertions
same about other

File: pom.xml (hunk 2 above)
Comment (mikhailtarasevich) [reply pom.xml]: +

File: src/test/java/CharCounterApplicationTest.java (hunk 1 above)
Comment (mikhailtarasevich) [reply src/test/java/CharCounterApplicationTest.java]: it seems like i used russian "c". edited.

File: pom.xml (hunk 3 above)
Comment (mikhailtarasevich) [reply pom.xml]: I used it at the beginning, but then I forgot to remove it. removed.

File: pom.xml (hunk 4 above)
Comment (mikhailtarasevich) [reply pom.xml:44]: +

File: pom.xml (hunk 5 above)
Comment (mikhailtarasevich) [reply pom.xml]: stackoverflow ctrl+c+v:(
deleted.

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java (hunk 11 above)
Comment (mikhailtarasevich) [reply src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: +

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java (hunk 13 above)
Comment (mikhailtarasevich) [reply src/test/java/com/mikhail/tarasevich/charcounter/provider/HashMapToStringConverterImplTest.java]: +

File: src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java (hunk 15 above)
Comment (mikhailtarasevich) [reply src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java]: rewrote for assertj methods.

File: src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+		this.validator = validator;
+		this.charCounter = charCounter;
+		this.viewProvider = viewProvider;
+		this.cacheProvider = cacheProvider;
+	}
+
+	public String countCharactersInText(String text) {
+		validator.validate(text);
+		if (cacheProvider.contains(text)) {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java]: remove empty line

File: src/main/java/com/mikhail/tarasevich/charcounter/ConsoleForCharCounterApplication.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+public class ConsoleForCharCounterApplication {
+
+	public static void main(String[] args) {
+		Validator validator = new ValidatorImpl();
+		CharCounter charCounter = new CharCounterImpl();
+		ViewProvider viewProvider = new ViewProviderImpl();
+		CacheProvider<String, LinkedHashMap<Character, Long>> cacheProvider = new CacheProviderImpl<>();
+
+		CharCounterApplication �harCounterApplication = new CharCounterApplication(validator, charCounter, viewProvider, cacheProvider);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/ConsoleForCharCounterApplication.java]: please use UTF-8 encoding

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class CacheProviderImpl<K,V> implements CacheProvider<K, V> {
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: please make this class not generic

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	private K key;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: where you use this field? Same about V

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+	private V value;
+	private Cache cache = new Cache(2);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: let`s inject size of cache in CacheProvider constructor

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+			return size() > size;
+		}
+	}
+
+	public void put(K key, V value) {
+		cache.put(key, value);
+	}
+
+	public V get(K key) {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: remove empty line

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+		cache.put(key, value);
+	}
+
+	public V get(K key) {
+		
+		return cache.get(key);
+	}
+
+	public boolean contains(K key) {
+
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: remove empty line

File: src/main/java/com/mikhail/tarasevich/charcounter/validator/Validator.java [hunk 23]
```diff
@@ -0,0 +1,5 @@
+package com.mikhail.tarasevich.charcounter.validator;
+
+public interface Validator {
+	void validate(String sentence);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/validator/Validator.java:4]: add empty line adound declaration of method, or remove this at ViewProvider and CharCounter intrefaces

File: src/main/java/com/mikhail/tarasevich/charcounter/validator/Validator.java [hunk 24]
```diff
@@ -0,0 +1,5 @@
...
+}
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/validator/Validator.java:5]: in the end of every file must be empty line 
https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

File: src/test/java/CharCounterApplicationTest.java [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+		LinkedHashMap<Character, Long> map1 = new LinkedHashMap<Character, Long>();
+		map1.put('h', 1L);
+		map1.put('i', 2L);
+
+		when(cacheProvider.contains(text)).thenReturn(false);
+		when(charCounter.uniqueCharCounter("hi")).thenReturn(map1);
+		when(viewProvider.provideView(map1, text)).thenReturn("Hi" + "\n\"H\" = 1;\n\"i\" = 1.");
+
+		assertEquals(charCounterApplication.countCharactersInText(text), "Hi" + "\n\"H\" = 1;\n\"i\" = 1.");
+	}
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java:58]: before the test is complete you have to make sure the mocks have been called

File: src/test/java/CharCounterApplicationTest.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+	private CharCounter charCounter;
+	@Mock
+	private ViewProvider viewProvider;
+	@Mock
+	private CacheProvider<String, LinkedHashMap<Character, Long>> cacheProvider;
+
+	private CharCounterApplication charCounterApplication;
+
+	@BeforeEach
+	void initMocks() {
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: use @ExtendWith agains this method

File: src/test/java/CharCounterApplicationTest.java [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+		MockitoAnnotations.initMocks(this);
+	}
+
+	@BeforeEach
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: why do you create new instance before every test?
use @InjectMocks

File: src/test/java/CharCounterApplicationTest.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+		String text = "hi";
+		LinkedHashMap<Character, Long> map1 = new LinkedHashMap<Character, Long>();
+		map1.put('h', 1L);
+		map1.put('i', 2L);
+
+		when(cacheProvider.contains(text)).thenReturn(true);
+		when(cacheProvider.get(text)).thenReturn(map1);
+		when(viewProvider.provideView(cacheProvider.get(text), text)).thenReturn("Hi" + "\n\"H\" = 1;\n\"i\" = 1.");
+
+		assertEquals(charCounterApplication.countCharactersInText(text), "Hi" + "\n\"H\" = 1;\n\"i\" = 1.");
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java:73]: before the test is complete you have to make sure the mocks have been called

File: src/test/java/CharCounterApplicationTest.java [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	@Test
+	void countCharactersInText_inputIsNullOrEmptyString_expectedException() {
+
+		String text = " ";
+
+		doThrow(new IllegalArgumentException()).when(validator).validate(text);
+
+		assertThrows(IllegalArgumentException.class, () -> {
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: remove braces, write in one line

File: src/test/java/CharCounterApplicationTest.java [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+			charCounterApplication.countCharactersInText(text);
+		});
```
Comment (AlexeyChirkov94) [inline src/test/java/CharCounterApplicationTest.java]: before the test is complete you have to make sure the mocks validator have been called, and over mocks was`n.
(verifyNoMoreInteractions)

File: src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+import com.mikhail.tarasevich.charcounter.provider.ViewProvider;
+import com.mikhail.tarasevich.charcounter.provider.CacheProvider;
+import com.mikhail.tarasevich.charcounter.validator.Validator;
+
+public class CharCounterApplication {
+	public LinkedHashMap<Character, Long> result;
+	private Validator validator;
+	private CharCounter charCounter;
+	private ViewProvider viewProvider;
+	private CacheProvider<String, LinkedHashMap<Character, Long>> cacheProvider;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java]: make all fields final  too

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProvider.java [hunk 32]
```diff
@@ -0,0 +1,10 @@
+package com.mikhail.tarasevich.charcounter.provider;
+
+public interface CacheProvider<K, V> {
+    
+	void put(K key, V value);
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProvider.java:4]: why double indent?d

File: src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.LinkedHashMap;
+
+import com.mikhail.tarasevich.charcounter.provider.CharCounter;
+import com.mikhail.tarasevich.charcounter.provider.ViewProvider;
+import com.mikhail.tarasevich.charcounter.provider.CacheProvider;
+import com.mikhail.tarasevich.charcounter.validator.Validator;
+
+public class CharCounterApplication {
+	public LinkedHashMap<Character, Long> result;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java]: this field not relate to class, and we never make fields public

File: src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Map;
+
+public class CacheProviderImpl<K,V> implements CacheProvider<K, V> {
+	private K key;
+	private V value;
+	private Cache cache = new Cache(2);
+
+	private class Cache extends LinkedHashMap<K, V> {
+
+		private int size;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImpl.java]: make it final too

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+
+class CharCounterImplTest {
+
+	CharCounterImpl charCounterImpl = new CharCounterImpl();
+	String text;
+	LinkedHashMap<Character, Long> expectedLinkedHashMap = new LinkedHashMap<Character, Long>();
+
+	@Test
+	void uniqueCharCounter_inputFromWordsSymbolsNumerals_expectedAmountOfUniqueCharacters() {
+
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: remove empty line

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+		expectedLinkedHashMap.put('H', 1l);
+		expectedLinkedHashMap.put('e', 1l);
+		expectedLinkedHashMap.put('l', 1l);
+		expectedLinkedHashMap.put(' ', 1l);
+		expectedLinkedHashMap.put('2', 2l);
+		text = "Hel 22";
+		assertEquals(expectedLinkedHashMap, charCounterImpl.uniqueCharCounter(text));
+
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: remove empty line

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+	@Test
+	void uniqueCharCounter_inputFromWordsSymbolsNumerals_expectedAmountOfUniqueCharacters() {
+
+		expectedLinkedHashMap.put('H', 1l);
+		expectedLinkedHashMap.put('e', 1l);
+		expectedLinkedHashMap.put('l', 1l);
+		expectedLinkedHashMap.put(' ', 1l);
+		expectedLinkedHashMap.put('2', 2l);
+		text = "Hel 22";
+		assertEquals(expectedLinkedHashMap, charCounterImpl.uniqueCharCounter(text));
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java:24]: add empty line before assertEquals

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+
+	}
+
+	@Test
+	void uniqueCharCounter_inputFromWhiteSpaces_expectedAmountOfWhiteSpaces() {
+
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: same about over tests

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 39]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	@Test
+	void uniqueCharCounter_inputIsNull_expectedNullPointerException() {
+
+		expectedLinkedHashMap.put(null, 1l);
+		text = null;
+		assertThrows(NullPointerException.class, () -> {
+			charCounterImpl.uniqueCharCounter(text);
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: remove braces, write in one line

File: src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java [hunk 40]
```diff
@@ -0,0 +1,34 @@
+package com.mikhail.tarasevich.charcounter.validator;
+
+import org.junit.jupiter.api.Test;
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;;
+
+class ValidatorImplTest {
+
+	ValidatorImpl validatorTest = new ValidatorImpl();
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/validator/ValidatorImplTest.java]: make it private and final 
same about over fileds in all classes

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java [hunk 41]
```diff
@@ -0,0 +1,47 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.util.LinkedHashMap;
+import org.junit.jupiter.api.Test;
+
+class CharCounterImplTest {
+
+	CharCounterImpl charCounterImpl = new CharCounterImpl();
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CharCounterImplTest.java]: private and final all fields, where field must not change in runtime, if field can be changes in runtime just private

File: src/test/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImplTest.java [hunk 42]
```diff
@@ -0,0 +1,130 @@
...
+		map1.put('4', 4L);
+
+		String str2 = "gggggggttttrrr";
+		LinkedHashMap<Character, Long> map2 = new LinkedHashMap<Character, Long>();
+		map2.put('g', 7L);
+		map2.put('t', 4L);
+		map2.put('r', 3L);
+
+		LinkedHashMap<String, LinkedHashMap<Character, Long>> expectedLinkedHashMap = new LinkedHashMap<String, LinkedHashMap<Character, Long>>(
+				2);
```
Comment (AlexeyChirkov94) [inline src/test/java/com/mikhail/tarasevich/charcounter/provider/CacheProviderImplTest.java]: write  in one line

**Review by AlexeyChirkov94 — APPROVED** [review]

File: src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java [hunk 43]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.util.LinkedHashMap;
+
+import com.mikhail.tarasevich.charcounter.provider.CharCounter;
+import com.mikhail.tarasevich.charcounter.provider.ViewProvider;
+import com.mikhail.tarasevich.charcounter.provider.CacheProvider;
+import com.mikhail.tarasevich.charcounter.validator.Validator;
+
+public class CharCounterApplication {
+	private LinkedHashMap<Character, Long> result;
```
Comment (AlexeyChirkov94) [inline src/main/java/com/mikhail/tarasevich/charcounter/CharCounterApplication.java]: this field not re;ate to class fields

**Pull request merged**